File: C++/4-Stack/Problems/2-MinStack/MinStack.cpp
#include <iostream>
#include <stack>
#include <vector>
#include <climits> // For INT_MAX and INT_MIN
#include <algorithm> // For min()

using namespace std;

// Approach 1: Using a pair (value, min) in the stack
// For every element, store the current minimum along with the element itself.
class MinStack1 {
private:
    stack<pair<int, int>> s; // Stack of pairs: (value, currentMin)

public:
    MinStack1() {
        // Constructor is empty
    }

    void push(int val) {
        if (s.empty()) {
            s.push({val, val}); // First element: min is the value itself
        } else {
            int currentMin = min(val, s.top().second); // Calculate the new minimum
            s.push({val, currentMin});             // Push the value and the minimum
        }
    }

    void pop() {
        s.pop(); // Simply pop the top element
    }

    int top() {
        return s.top().first; // Return the value at the top
    }

    int getMin() {
        return s.top().second; // Return the minimum value at the top
    }
};

// Approach 2: Using two stacks (one for values, one for minimums)
// One stack stores all values, the other stores the minimums encountered so far.
class MinStack2 {
private:
    stack<int> s;      // Stack for storing the values
    stack<int> minStack; // Stack for storing the minimums

public:
    MinStack2() {
        // Constructor is empty
    }

    void push(int val) {
        s.push(val); // Push the value onto the value stack
        if (minStack.empty() || val <= minStack.top()) {
            minStack.push(val); // Push to minStack if it's empty or the new value is <= current min
        }
    }

    void pop() {
        if (s.top() == minStack.top()) {
            minStack.pop(); // Pop from minStack if the top value is the current minimum
        }
        s.pop(); // Always pop from the value stack
    }

    int top() {
        return s.top(); // Return the top value
    }

    int getMin() {
        return minStack.top(); // Return the top of the minimum stack
    }
};

// Approach 3: Using a single stack and encoding the min value
//  Clever approach: When a new min is found, store the *difference* between
//  the new min and the *previous* min, rather than the min itself.
class MinStack3 {
private:
    stack<long long> s;
    long long minVal; // Keep track of the global minimum

public:
    MinStack3() {
        minVal = INT_MAX;
    }

    void push(int val) {
        if (s.empty()) {
            minVal = val;
            s.push(val);
        } else {
            if (val < minVal) {
                s.push(2LL * val - minVal); // Store the difference (encoded min)
                minVal = val;             // Update the global minimum
            } else {
                s.push(val); // If not a new min, store the original value
            }
        }
    }

    void pop() {
        if (s.empty()) return;

        long long top = s.top();
        s.pop();

        if (top < minVal) {
            // If the top was an encoded min, update minVal
            minVal = 2 * minVal - top;
        }
    }

    int top() {
        if (s.empty()) return -1; // Or throw an exception
        long long top = s.top();
        if (top < minVal) {
            return (int)minVal; // return the minVal
        }
        return (int)top;
    }

    int getMin() {
        return (int)minVal;
    }
};

// Approach 4: Using a vector and a min variable.
// This approach uses a vector to store the elements and a separate variable to store the minimum value.
class MinStack4 {
private:
    vector<int> data;
    int minVal;

public:
    MinStack4() {
        minVal = INT_MAX;
    }

    void push(int val) {
        data.push_back(val);
        if (val < minVal) {
            minVal = val;
        }
    }

    void pop() {
        if (data.empty()) return;
        int poppedVal = data.back();
        data.pop_back();
        if (poppedVal == minVal) {
            // Need to find the new minimum
            minVal = INT_MAX; // Reset to max, and find min again.
            for (int val : data) {
                minVal = min(minVal, val);
            }
            if(data.empty()) {
               minVal = INT_MAX;
            }
        }
    }

    int top() {
        if (data.empty()) return -1;
        return data.back();
    }

    int getMin() {
        return minVal;
    }
};

// Approach 5: Using a linked list and a min variable.
// This approach uses a linked list to store the elements and a separate variable to store the minimum value.
class MinStack5 {
private:
    struct Node {
        int val;
        Node* next;
        int minVal; // Minimum value seen so far *including* this node
        Node(int x, int min) : val(x), next(nullptr), minVal(min) {}
    };
    Node* head;
    //int minVal; // Not needed, its in the Node.

public:
    MinStack5() : head(nullptr) {
        //minVal = INT_MAX;
    }

    void push(int val) {
        if (!head) {
            head = new Node(val, val);
            //minVal = val;
        } else {
            int currentMin = min(val, head->minVal);
            head = new Node(val, currentMin);
            head->next = head;
            head = head->next;
            head = new Node(val, currentMin); // Corrected: Create a new node, insert at the head.
            head->next = head->next;       // Make the new node point to the *old* head
            head = head->next;                 // Update head to the new node.
        }
    }

    void pop() {
        if (!head) return;
        Node* temp = head;
        head = head->next;
        delete temp;
    }

    int top() {
        if (!head) return -1;
        return head->val;
    }

    int getMin() {
        if (!head) return INT_MAX;
        return head->minVal;
    }
};

int main() {
    // Example usage of each MinStack implementation:
    cout << "Approach 1: Using a pair (value, min) in the stack" << endl;
    MinStack1 ms1;
    ms1.push(3);
    ms1.push(5);
    ms1.push(2);
    ms1.push(1);
    cout << "Top: " << ms1.top() << endl;       // Output: 1
    cout << "Min: " << ms1.getMin() << endl; // Output: 1
    ms1.pop();
    cout << "Top: " << ms1.top() << endl;       // Output: 2
    cout << "Min: " << ms1.getMin() << endl; // Output: 2

    cout << "\nApproach 2: Using two stacks (one for values, one for minimums)" << endl;
    MinStack2 ms2;
    ms2.push(3);
    ms2.push(5);
    ms2.push(2);
    ms2.push(1);
    cout << "Top: " << ms2.top() << endl;       // Output: 1
    cout << "Min: " << ms2.getMin() << endl; // Output: 1
    ms2.pop();
    cout << "Top: " << ms2.top() << endl;       // Output: 2
    cout << "Min: " << ms2.getMin() << endl; // Output: 2

    cout << "\nApproach 3: Using a single stack and encoding the min value" << endl;
    MinStack3 ms3;
    ms3.push(3);
    ms3.push(5);
    ms3.push(2);
    ms3.push(1);
    cout << "Top: " << ms3.top() << endl;       // Output: 1
    cout << "Min: " << ms3.getMin() << endl; // Output: 1
    ms3.pop();
    cout << "Top: " << ms3.top() << endl;       // Output: 2
    cout << "Min: " << ms3.getMin() << endl; // Output: 2

    cout << "\nApproach 4: Using a vector and a min variable." << endl;
    MinStack4 ms4;
    ms4.push(3);
    ms4.push(5);
    ms4.push(2);
    ms4.push(1);
    cout << "Top: " << ms4.top() << endl;       // Output: 1
    cout << "Min: " << ms4.getMin() << endl; // Output: 1
    ms4.pop();
    cout << "Top: " << ms4.top() << endl;       // Output: 2
    cout << "Min: " << ms4.getMin() << endl; // Output: 2

    cout << "\nApproach 5: Using a linked list and a min variable." << endl;
    MinStack5 ms5;
    ms5.push(3);
    ms5.push(5);
    ms5.push(2);
    ms5.push(1);
    cout << "Top: " << ms5.top() << endl;
    cout << "Min: " << ms5.getMin() << endl;
    ms5.pop();
    cout << "Top: " << ms5.top() << endl;
    cout << "Min: " << ms5.getMin() << endl;

    return 0;
}

File: Java/4-Stack/Problems/2-MinStack/MinStack.java
import java.util.Stack;

public class MinStack {

    // Approach 1: Using a pair (value, min) in the stack
    // Each element in the stack keeps track of the minimum value seen so far.
    static class MinStack1 {
        private Stack<Pair> stack;

        public MinStack1() {
            stack = new Stack<>();
        }

        public void push(int val) {
            int min = val;
            if (!stack.isEmpty()) {
                min = Math.min(min, stack.peek().min);
            }
            stack.push(new Pair(val, min));
        }

        public void pop() {
            stack.pop();
        }

        public int top() {
            return stack.peek().val;
        }

        public int getMin() {
            return stack.peek().min;
        }

        private static class Pair {
            int val;
            int min;

            public Pair(int val, int min) {
                this.val = val;
                this.min = min;
            }
        }
    }

    // Approach 2: Using two stacks (one for values, one for minimums)
    // One stack stores the values, and the other stores the minimums seen so far.
    static class MinStack2 {
        private Stack<Integer> stack;
        private Stack<Integer> minStack;

        public MinStack2() {
            stack = new Stack<>();
            minStack = new Stack<>();
        }

        public void push(int val) {
            stack.push(val);
            if (minStack.isEmpty() || val <= minStack.peek()) {
                minStack.push(val);
            }
        }

        public void pop() {
            int popped = stack.pop();
            if (!minStack.isEmpty() && popped == minStack.peek()) {
                minStack.pop();
            }
        }

        public int top() {
            return stack.peek();
        }

        public int getMin() {
            return minStack.peek();
        }
    }

    // Approach 3: Optimized two stacks (one for values, one for minimums with duplicates)
    //  The minStack only pushes when a new minimum is found
    static class MinStack3 {
        private Stack<Integer> stack;
        private Stack<Integer> minStack;

        public MinStack3() {
            stack = new Stack<>();
            minStack = new Stack<>();
        }

        public void push(int val) {
            stack.push(val);
            if (minStack.isEmpty() || val < minStack.peek()) { // Changed to '<'
                minStack.push(val);
            } else if (val == minStack.peek()) {
                minStack.push(val); //push duplicates
            }
        }

        public void pop() {
            int popped = stack.pop();
            if (!minStack.isEmpty() && popped == minStack.peek()) {
                minStack.pop();
            }
        }

        public int top() {
            return stack.peek();
        }

        public int getMin() {
            return minStack.peek();
        }
    }

    // Approach 4: Using a single stack and encoding min values
    // This approach uses a single stack, but encodes the current minimum value
    // along with the actual value.  This is a space optimization.
    static class MinStack4 {
        private Stack<Long> stack;
        private long min;

        public MinStack4() {
            stack = new Stack<>();
        }

        public void push(int val) {
            if (stack.isEmpty()) {
                min = val;
                stack.push((long)val); // Store directly as long
            } else {
                if (val < min) {
                    stack.push((long)val * 2 - min);  // Store encoded value
                    min = val;
                } else {
                    stack.push((long)val);
                }
            }
        }

        public void pop() {
            if (!stack.isEmpty()) {
                long top = stack.pop();
                if (top < min) {
                    min = 2 * min - top;  // Decode the minimum value.
                }
            }
        }

        public int top() {
            if (!stack.isEmpty()) {
                long top = stack.peek();
                if (top < min) {
                    return (int)min; // Return current min
                } else {
                    return (int)top;
                }
            }
            return -1; // Or throw an exception
        }

        public int getMin() {
            return (int)min;
        }
    }

    // Approach 5: Using a linked list with min tracking.
    // This approach uses a custom linked list node that also stores the minimum
    // value seen so far.
    static class MinStack5 {

        private static class Node {
            int val;
            int min;
            Node next;

            public Node(int val, int min, Node next) {
                this.val = val;
                this.min = min;
                this.next = next;
            }
        }

        private Node head;

        public MinStack5() {
            head = null;
        }

        public void push(int val) {
            int min = val;
            if (head != null) {
                min = Math.min(min, head.min);
            }
            head = new Node(val, min, head);
        }

        public void pop() {
            if (head != null) {
                head = head.next;
            }
        }

        public int top() {
            if (head != null) {
                return head.val;
            }
            return -1; // Or throw an exception.
        }

        public int getMin() {
            if (head != null) {
                return head.min;
            }
            return -1; // Or throw an exception.
        }
    }

    public static void main(String[] args) {
        // Example Usage for each MinStack implementation.  Added more tests.
        System.out.println("--- MinStack1 (Pair) ---");
        MinStack1 ms1 = new MinStack1();
        ms1.push(3);
        ms1.push(5);
        ms1.push(2);
        ms1.push(1);
        System.out.println("Top: " + ms1.top()); // 1
        System.out.println("Min: " + ms1.getMin()); // 1
        ms1.pop();
        System.out.println("Top: " + ms1.top()); // 2
        System.out.println("Min: " + ms1.getMin()); // 2
        ms1.pop();
        System.out.println("Min: " + ms1.getMin()); // 3

        System.out.println("\n--- MinStack2 (Two Stacks) ---");
        MinStack2 ms2 = new MinStack2();
        ms2.push(3);
        ms2.push(5);
        ms2.push(2);
        ms2.push(1);
        System.out.println("Top: " + ms2.top());
        System.out.println("Min: " + ms2.getMin());
        ms2.pop();
        System.out.println("Top: " + ms2.top());
        System.out.println("Min: " + ms2.getMin());
        ms2.pop();
        System.out.println("Min: " + ms2.getMin());

        System.out.println("\n--- MinStack3 (Two Stacks Optimized) ---");
        MinStack3 ms3 = new MinStack3();
        ms3.push(3);
        ms3.push(5);
        ms3.push(2);
        ms3.push(1);
        ms3.push(1); //push duplicate min
        System.out.println("Top: " + ms3.top());
        System.out.println("Min: " + ms3.getMin());
        ms3.pop();
        System.out.println("Top: " + ms3.top());
        System.out.println("Min: " + ms3.getMin());
        ms3.pop();
        System.out.println("Min: " + ms3.getMin());

        System.out.println("\n--- MinStack4 (Single Stack Encoded) ---");
        MinStack4 ms4 = new MinStack4();
        ms4.push(3);
        ms4.push(5);
        ms4.push(2);
        ms4.push(1);
        System.out.println("Top: " + ms4.top());
        System.out.println("Min: " + ms4.getMin());
        ms4.pop();
        System.out.println("Top: " + ms4.top());
        System.out.println("Min: " + ms4.getMin());
        ms4.pop();
        System.out.println("Min: " + ms4.getMin());

        System.out.println("\n--- MinStack5 (Linked List) ---");
        MinStack5 ms5 = new MinStack5();
        ms5.push(3);
        ms5.push(5);
        ms5.push(2);
        ms5.push(1);
        System.out.println("Top: " + ms5.top());
        System.out.println("Min: " + ms5.getMin());
        ms5.pop();
        System.out.println("Top: " + ms5.top());
        System.out.println("Min: " + ms5.getMin());
        ms5.pop();
        System.out.println("Min: " + ms5.getMin());
    }
}

